faunadb-scala/src/main/scala/faunadb/values/Value.scala

```diff
@@ -26,245 +23,161 @@ import scala.annotation.meta.{param, field, getter}
-    */
-  def asDateOpt: Option[Date] = None
-  def asDate = asDateOpt.get
+/**
+  *  A Boolean value.
+  *
+  * '''Reference''': [[https://faunadb.com/documentation/queries#values FaunaDB Values]]
+  */
+object BooleanV {
+  def apply(b: Boolean) = if (b) TrueV else FalseV
+  def unapply(b: BooleanV) = b
+}
+sealed abstract class BooleanV(@(JsonValue @getter) val value: Boolean) extends ScalarValue {
+  // satisfy name-based extractor interface
+  val isEmpty = false
+  val get = value
+}
+case object TrueV extends BooleanV(true)
+case object FalseV extends BooleanV(false)
 
-  /**
-   * Accesses the value of the specified field if this is an object node.
-   */
-  def apply(key: String): Value = asObject(key)
+// Fauna special types
 
-  /**
-   * Accesses the value of the specified element if this is an array node.
-   */
-  def apply(index: Int): Value = asArray(index)
+/**
+  * A Ref.
+  *
+  * '''Reference''': [[https://faunadb.com/documentation/queries#values-special_types FaunaDB Special Types]]
+  */
+object Ref {
+  def apply(clss: Ref, id: String): Ref = Ref(s"${clss.value}/$id")
+}
 
-  /**
-   * Accesses the value of the specified field if this is an object node.
-   */
-  def get(key: String): Option[Value] = asObjectOpt.flatMap(_.get(key))
+case class Ref(@(JsonProperty @field @param)("@ref") value: String) extends ScalarValue
 
-  /**
-   * Accesses the value of the specified element if this is an array node.
-   */
-  def get(index: Int): Option[Value] = asArrayOpt.flatMap(_.lift(index))
+/**
+  * A Set Ref.
+  *
+  * '''Reference''': [[https://faunadb.com/documentation/queries#values-special_types FaunaDB Special Types]]
+  */
+case class SetRef(@JsonProperty("@set") parameters: Value) extends ScalarValue
 
+/**
+  * A Timestamp.
+  *
+  * '''Reference''': [[https://faunadb.com/documentation/queries#values-special_types FaunaDB Special Types]]
+  */
+object Timestamp {
+  def apply(value: String): Timestamp =
+    Timestamp(ZonedDateTime.parse(value, DateTimeFormatter.ISO_OFFSET_DATE_TIME).toInstant)
 }
 
-case class RawV(@(JsonValue @getter) value: JsonNode) extends Value
-
-case class StringV(@(JsonValue @getter) value: String) extends Value {
-  override def asStringOpt: Option[String] = Some(value)
+case class Timestamp(@(JsonIgnore @param @field @getter) instant: Instant) extends ScalarValue {
+  @JsonProperty("@ts")
+  val strValue = instant.toString
 }
 
-case class NumberV(@(JsonValue @getter) value: Long) extends Value {
-  override def asNumberOpt: Option[Long] = Some(value)
+/**
+  * A Date.
+  *
+  * '''Reference''': [[https://faunadb.com/documentation/queries#values-special_types FaunaDB Special Types]]
+  */
+object Date {
+  def apply(value: String): Date = Date(LocalDate.parse(value))
 }
 
-case class DoubleV(@(JsonValue @getter) value: Double) extends Value {
-  override def asDoubleOpt: Option[Double] = Some(value)
+case class Date(@(JsonIgnore @param @field @getter) localDate: LocalDate) extends ScalarValue {
```

Comment:
"local" dates merit some consideration, as the date on either side of the network may not be the same.